iree/compiler/Dialect/HAL/Transforms/PackAllocations.cpp

```diff
@@ -0,0 +1,367 @@
+                                        BufferConstraintsAttr bufferConstraints,
+                                        OpBuilder &builder) {
+    auto loc = packOp.getLoc();
+    int64_t offsetAlignment =
+        bufferConstraints.min_buffer_offset_alignment().getSExtValue();
+    int64_t rangeAlignment =
+        bufferConstraints.min_buffer_range_alignment().getSExtValue();
+
+    // Bucket all slices by their size SSA value. We rely on shapes being
+    // lowered to computed byte sizes and CSE to then dedupe the values for us.
+    llvm::MapVector<Value, SmallVector<const Slice *>> slicesBySize;
+    for (auto &slice : slices) {
+      slicesBySize[slice.dynamicSize].push_back(&slice);
+    }
+
+    // Allocate each bucket while observing the lifetime of the slices within.
+    // Two or more slices may overlap in lifetime and need their own unique
+    // reservation.
+    Value offset = baseOffset;
+    for (auto &sizeBucket : slicesBySize) {
+      auto sliceSize = align(loc, sizeBucket.first, rangeAlignment, builder);
+      auto &slices = sizeBucket.second;
+      std::stable_sort(slices.begin(), slices.end());
+
+      // Bin the slices by those who do not overlap. All of the allocations in
+      // each bin can alias.
+      // NOTE: O(n^2) in the worst case but there's usually only a small number
+      // of bins (<10) as we have already bucketed by size class. We could do
+      // some sorting and make this O(nlogn) or O(logn) with some interval tree
+      // magic.
+      struct Bin {
+        Value offset;
+        SmallVector<const Slice *> slices;
+        bool intersects(const Slice &slice) const {
+          for (auto *binSlice : slices) {
+            if (binSlice->intersects(slice)) return true;
+          }
+          return false;
+        }
+      };
+      SmallVector<Bin> bins;
+      for (auto *slice : slices) {
+        // Try to find a bin we can reuse (non-intersecting lifetime).
+        Bin *targetBin = nullptr;
+        for (auto &bin : bins) {
+          if (!bin.intersects(*slice)) {
+            targetBin = &bin;
+            break;
+          }
+        }
+        if (!targetBin) {
+          // Allocate a new bin for this slice.
+          bins.push_back({offset, {}});
+          targetBin = &bins.back();
+          offset =
+              align(loc, builder.createOrFold<AddIOp>(loc, offset, sliceSize),
+                    offsetAlignment, builder);
+        }
+        targetBin->slices.push_back(slice);
+        slice->packedOffset.replaceAllUsesWith(targetBin->offset);
+      }
+    }
+
+    return align(loc, offset, rangeAlignment, builder);
+  }
+
+  TargetOptions targetOptions_;
+};
+
+std::unique_ptr<OperationPass<FuncOp>> createPackAllocationsPass(
+    TargetOptions targetOptions) {
+  return std::make_unique<PackAllocationsPass>(targetOptions);
+}
+
+static PassRegistration<PackAllocationsPass> pass(
+    "iree-hal-pack-allocations",
```

Comment:
Where did you land on adding pass statistics to this pass?

Reply:
until we have them turned on I'm not bothering - easy to add later on when they are useful